XenAdmin/SplashScreenContext.cs

```diff
@@ -0,0 +1,120 @@
+﻿/* Copyright (c) Cloud Software Group Holdings, Inc.
+ * All rights reserved. 
+ * 
+ * Redistribution and use in source and binary forms, 
+ * with or without modification, are permitted provided 
+ * that the following conditions are met: 
+ * 
+ * *   Redistributions of source code must retain the above 
+ *     copyright notice, this list of conditions and the 
+ *     following disclaimer. 
+ * *   Redistributions in binary form must reproduce the above 
+ *     copyright notice, this list of conditions and the 
+ *     following disclaimer in the documentation and/or other 
+ *     materials provided with the distribution. 
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND 
+ * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
+ * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR 
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGE.
+ */
+
+using System;
+using System.Configuration;
+using System.Reflection;
+using System.Threading;
+using System.Windows.Forms;
+using XenAdmin.Dialogs;
+
+namespace XenAdmin
+{
+    internal class SplashScreenContext : ApplicationContext
+    {
+        private static readonly log4net.ILog _log = log4net.LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);
+
+        private readonly SplashScreen _splashScreen;
+        private MainWindow _mainWindow;
+        private readonly string[] _args;
+
+        public static MainWindow MainWindow { get; private set; }
+
+        public SplashScreenContext(params string[] args)
+        {
+            _args = args;
+            _splashScreen = new SplashScreen();
+            _splashScreen.ShowMainWindowRequested += _splashScreen_ShowMainWindowRequested;
+            StartLoadingSettings();
+            _splashScreen.Show();
+        }
+
+        private void StartLoadingSettings()
+        {
+            ThreadPool.QueueUserWorkItem(obj =>
+            {
+                try
+                {
+                    Settings.Load();
+                    Settings.ReconfigureProxyAuthenticationSettings();
+                    Settings.ConfigureExternalSshClientSettings();
+                    Settings.Log();
+
+                    _splashScreen.AllowToClose = true;

```

Comment:
wouldn't it be better to get rid of the timer and field, and have a method named `StartMainWindow` or similar?

Reply:
The reason why I added both is that I wanted the SplashScreen to close after at least 2 seconds have elapsed or the settings have been loaded, whichever first. I couldn't think of a simpler way to do it.

Reply:
Quick idea, haven't tested or fleshed it out: what if we have an event in Settings, fired when they have completed loading. Then we can close the splash screen when that event is fired.

Reply:
Hmmm, loading entails calling all four of these methods (2 of which may additionally try a save operation which may last a bit), where would the event be fired?

Reply:
Plus that if loading settings happens to finish too quickly, I still think we should give the full 2 seconds before showing the main window, otherwise the splash won't show for too long.

Reply:
tried to flesh out my idea out but it but turned out to be more complex than the current implementation. This works 👍